File: src/models/Medicine.js
/**
 * Medicine Database Model Schema
 *
 * This is a reference model showing the relationships between Medicine, Company, and Distributor entities.
 * The actual implementation depends on the backend database system (SQL, NoSQL, etc.)
 */

/**
 * Medicine Schema
 */
const MedicineSchema = {
  id: "Number/UUID (Primary Key)",
  medicineId: "String (Business ID, unique)",
  name: "String (required)",
  category: "String (required)",
  description: "String",
  manufacturer: "String",
  batchNumber: "String",
  expiryDate: "Date",
  stock: "Number (required)",
  stockStatus: "String (High/Normal/Low/Critical)",
  price: "Number (required)",
  requiresPrescription: "Boolean",
  dosageForm: "String (Tablet/Capsule/Syrup/etc.)",
  strength: "String",
  interactions: "Array of Strings",
  sideEffects: "Array of Strings",

  // Relationships
  companyId: "Foreign Key to Company", // Many medicines can come from one company
  company: "Object (Reference to Company entity)",

  distributorId: "Foreign Key to Distributor", // Many medicines can be distributed by one distributor
  distributor: "Object (Reference to Distributor entity)",

  // Timestamps
  createdAt: "Date",
  updatedAt: "Date",
};

/**
 * Company Schema
 */
const CompanySchema = {
  id: "Number/UUID (Primary Key)",
  companyId: "String (Business ID, unique)",
  name: "String (required)",
  address: "String (required)",
  phone: "String (required)",
  email: "String",
  website: "String",
  contactPerson: "String",
  licenseNumber: "String",
  registrationDate: "Date",

  // Relationships
  medicines: "Array of Medicine references", // One-to-Many: One company produces many medicines

  // Timestamps
  createdAt: "Date",
  updatedAt: "Date",
};

/**
 * Distributor Schema
 */
const DistributorSchema = {
  id: "Number/UUID (Primary Key)",
  distributorId: "String (Business ID, unique)",
  name: "String (required)",
  logoUrl: "String",
  region: "String (required)",
  headquarters: "String",
  areas: "Array of Strings",
  contactName: "String",
  contactTitle: "String",
  phone: "String (required)",
  email: "String",
  website: "String",
  relationshipSince: "Date",
  contractStatus: "String (Active/Pending/Expired/On Hold)",
  contractRenewal: "Date",
  deliveryTime: "String",
  rating: "Number",
  reliability: "String",
  lastDelivery: "Date",
  paymentTerms: "String",
  specialties: "Array of Strings",

  // Relationships
  medicines: "Array of Medicine references", // One-to-Many: One distributor distributes many medicines

  // Timestamps
  createdAt: "Date",
  updatedAt: "Date",
};

/**
 * Relationship Summary:
 *
 * 1. Medicine to Company: Many-to-One
 *    - A medicine belongs to only one company
 *    - A company can produce many medicines
 *
 * 2. Medicine to Distributor: Many-to-One
 *    - A medicine is distributed by one distributor
 *    - A distributor can distribute many medicines
 *
 * Note: In a more complex system, the Medicine-to-Distributor relationship
 * could be Many-to-Many, where a medicine could be distributed by multiple
 * distributors and a distributor handles many medicines. That would require
 * a junction table (MedicineDistributor) to manage the relationship.
 */

File: backend/src/main/java/com/hospital/pharmacy/model/Medicine.java
package com.hospital.pharmacy.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "medicines")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Medicine {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Column(unique = true)
    private String medicineId;  // Business ID like "MED-001"

    @NotBlank
    private String name;

    @NotBlank
    private String category;

    @Column(length = 1000)
    private String description;

    private String manufacturer;

    private String batchNumber;

    @NotNull
    private LocalDate expiryDate;

    @NotNull
    @Positive
    private Integer stock;

    private String stockStatus;  // "High", "Normal", "Low", "Critical"

    @NotNull
    @Positive
    private BigDecimal price;

    private boolean requiresPrescription;

    private String dosageForm;  // "Tablet", "Capsule", "Syrup", etc.

    private String strength;  // "10mg", "500mg", etc.

    @ElementCollection
    @CollectionTable(name = "medicine_interactions", joinColumns = @JoinColumn(name = "medicine_id"))
    @Column(name = "interaction")
    private List<String> interactions = new ArrayList<>();

    @ElementCollection
    @CollectionTable(name = "medicine_side_effects", joinColumns = @JoinColumn(name = "medicine_id"))
    @Column(name = "side_effect")
    private List<String> sideEffects = new ArrayList<>();

    @ManyToOne
    @JoinColumn(name = "company_id")
    private Company company;

    @ManyToOne
    @JoinColumn(name = "distributor_id")
    private Distributor distributor;
} 